lru.go

```diff
@@ -4,116 +4,191 @@
-	}
-	c := &Cache{
-		size:      size,
-		evictList: list.New(),
-		items:     make(map[interface{}]*list.Element, size),
-	}
-	return c, nil
+// New creates an LRU cache. It accepts a maximum size limit
+// and a maximum byte limit. If either one of these are set 
+// to zero, then that limit will be considered unlimitted but 
+// if both are set to zero then an error will be returned.
+// Exceeding either of these limits while calling Add() will
+// result in the least recently used item being dropped from
+// the cache and the onRemove() function being called for
+// that item.
+func New(items int, bytes int64) (*Cache, error) {
+    if items <= 0 && bytes <= int64(0) {
+        return nil, errors.New("Must provide a positive bound on either item count or byte count.")
+    }
+    onRemove := func(val interface{}) {
+        return
+    }
+    c := &Cache{
+        maxBytes:   bytes,
+        bytes:      int64(0),
+        maxItems:   items,
+        evictList:  list.New(),
+        items:      make(map[interface{}]*list.Element),
+        onRemove:   onRemove,
+    }
+    return c, nil
 }
 
-// Purge is used to completely clear the cache
+// Purge will completely clear the cache of all key, value pairs.
 func (c *Cache) Purge() {
-	c.lock.Lock()
-	defer c.lock.Unlock()
-	c.evictList = list.New()
-	c.items = make(map[interface{}]*list.Element, c.size)
+    c.lock.Lock()
+    defer c.lock.Unlock()            
+    for c.Len() > 0 {
+        c.removeOldest()
+    }
+    c.items = nil
+    c.items = make(map[interface{}]*list.Element)
 }
 
-// Add adds a value to the cache.
+// Schedule routine purges. May be used to free up memory or 
+// ensure that the OnRemove function is called within a certain
+// amount of time.
+func (cache *Cache) ScheduleClear(d time.Duration) {
+    t := time.Tick(d)
+    for {
+        select {
+        case <-t:
+            cache.Purge()
+        }
+    }
+}
+
+// Add adds a value to the cache. If the new value causes the
+// set maxItems or maxBytes values to be exceeded, then the
+// least recently used item will be popped from the cache and 
+// the onRemove() function will be called for that value.
 func (c *Cache) Add(key, value interface{}) {
-	c.lock.Lock()
-	defer c.lock.Unlock()
-
-	// Check for existing item
-	if ent, ok := c.items[key]; ok {
-		c.evictList.MoveToFront(ent)
-		ent.Value.(*entry).value = value
-		return
-	}
-
-	// Add new item
-	ent := &entry{key, value}
-	entry := c.evictList.PushFront(ent)
-	c.items[key] = entry
-
-	// Verify size not exceeded
-	if c.evictList.Len() > c.size {
-		c.removeOldest()
-	}
+    c.lock.Lock()
+    defer c.lock.Unlock()
+
+    // Check for existing item
+    if ent, ok := c.items[key]; ok {
+        c.evictList.MoveToFront(ent)
+        ent.Value.(*entry).value = value
+        return
+    }
+
+    // Add new item
+    ent := &entry{key, value}
+    entry := c.evictList.PushFront(ent)
+    c.items[key] = entry
+
+    buf := value.([]byte)
```

Comment:
Good call, as I said I made a lot of these changes to satisfy my own needs but it would make more sense to make it universally usable. I went ahead and updated with a method that gets the byte count by gob encoding and getting the length of that. It might slow down the processing a little though, so it is a bit of a tradeoff.